File: code401challenges/src/main/java/code401challenges/stacksAndQueues/Stack.java
package code401challenges.stacksAndQueues;

import java.util.NoSuchElementException;

public class Stack {
    //instance variables
    Node top;

    //constructor function
    Stack () {
        this.top = null;
    }

    //instance methods
    public void push (String nodeValue) {
        Node node = new Node(nodeValue);
        node.next = this.top;
        this.top = node;
    }

    public String pop () throws NoSuchElementException {
        if (this.top == null) {
            throw new NoSuchElementException("Cannot perform pop() on empty stack");
        }
        String value = this.top.value;
        this.top = this.top.next;
        return value;
    }

    public String peek () throws NoSuchElementException{
        if (this.top != null) {
            return this.top.value;
        } else {
            throw new NoSuchElementException("Cannot perform peek() on empty stack");
        }
    }

    public Boolean isEmpty () {
        return this.top == null ? true : false;
    }


}

File: js401challenges/Stacks_Queues/Stack.js
'use strict';

const Node = require('./Node.js')

class Stack {

  constructor () {
    this.top = 0;
  }

  push (value) {
    if (this.isEmpty) {
      let node = new Node(value)
      this[this.top++] = node;
    } else {
      let node = new Node(value, this[this.top]);
      this[this.top++] = node;
    }
  }

  pop () {
    if (this.top === 0) throw new Error('Cannot pop from empty stack');
    let poppedNode = this[this.top];
    delete this[this.top];
    this.top--;
    return poppedNode;
  }

  peek () {
    if (this.top === 0) throw new Error('Cannot pop from empty stack');
    return this[this.top].value;
  }

  isEmpty () {
    return (this.top === 0) ? true : false;
  }

}

module.exports = Stack;
